plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py

```diff
@@ -0,0 +1,180 @@
+
+    Returns:
+    - str: The path to the downloaded file.
+    """
+
+    if not url.startswith("http"):
+        raise ValueError(f"URL {url} is not valid. Only http/https is supported.")
+
+    # Derive the local filename from the URL
+    local_file_name = os.path.join(target_dir, os.path.basename(url))
+
+    fs = fsspec.filesystem("http")
+
+    # Use fsspec to get the remote file and save it locally
+    logger.info(f"Downloading {url}... to {os.path.abspath(local_file_name)}")
+    fs.get(url, local_file_name)
+    logger.info("File downloaded successfully!")
+
+    return local_file_name
+
+
+def download_vscode(
+    code_server_remote_path: str,
+    code_server_dir_name: str,
+):
+    """
+    Download vscode server and plugins from remote to local and add the directory of binary executable to $PATH.
+
+    Parameters:
+    - code_server_remote_path (str): The URL of the code-server tarball.
+    - code_server_dir_name (str): The name of the code-server directory.
+    Return:
+    - str: The path to the code-server binary.
+    """
+
+    # If the code server already exists in the container, skip downloading
+    executable_path = shutil.which(EXECUTABLE_NAME)
+    if executable_path is not None:
+        logger.info(f"Code server binary already exists at {executable_path}")
+        logger.info("Skipping downloading code server...")
+        return
+
+    logger.info("Code server is not in $PATH, start downloading code server...")
+
+    # Create DOWNLOAD_DIR if not exist
+    logger.info(f"DOWNLOAD_DIR: {DOWNLOAD_DIR}")
+    os.makedirs(DOWNLOAD_DIR, exist_ok=True)
+
+    logger.info(f"Start downloading files to {DOWNLOAD_DIR}")
+
+    # Download remote file to local
+    code_server_tar_path = download_file(code_server_remote_path, DOWNLOAD_DIR)
+
+    # Extract the tarball
+    with tarfile.open(code_server_tar_path, "r:gz") as tar:
+        tar.extractall(path=DOWNLOAD_DIR)
+
+    code_server_dir_path = os.path.join(DOWNLOAD_DIR, code_server_dir_name)
+
+    code_server_bin_dir = os.path.join(code_server_dir_path, "bin")
+
+    # Add the directory of code-server binary to $PATH
+    os.environ["PATH"] = code_server_bin_dir + os.pathsep + os.environ["PATH"]
+
+
+def vscode(
+    _task_function: Optional[Callable] = None,
+    server_up_seconds: Optional[int] = DEFAULT_UP_SECONDS,
+    port: Optional[int] = 8080,
+    enable: Optional[bool] = True,
+    code_server_remote_path: Optional[
+        str
+    ] = "https://github.com/coder/code-server/releases/download/v4.18.0/code-server-4.18.0-linux-amd64.tar.gz",
+    # The untarred directory name may be different from the tarball name
+    code_server_dir_name: Optional[str] = "code-server-4.18.0-linux-amd64",
+    pre_execute: Optional[Callable] = None,
+    post_execute: Optional[Callable] = None,
+):
+    """
+    vscode decorator modifies a container to run a VSCode server:
```

Comment:
Add this vscode behavior to PR description for reviewer to easily understand the context